lib/pbench/cli/agent/commands/tools/clear.py

```diff
@@ -25,83 +25,104 @@ def __init__(self, context):
     def execute(self) -> int:
         errors = 0
 
-        if self.verify_tool_group(self.context.group) != 0:
-            return 1
-
-        try:
-            remotes = self.context.remote.split(",")
-        except Exception:
-            # We were not given any remotes on the command line, build the list
-            # from the tools group directory.
-            remotes = self.remote(self.tool_group_dir)
-            if not remotes:
-                # FIXME:  This is a weird case -- the group directory exists,
-                #  but contains no remotes. This is the result of clearing the
-                #  tools from a group which has already been cleared, because
-                #  the first clear does not remove the group directory.  This
-                #  behavior might be desirable for the `default` directory, but
-                #  it seems wrong for custom groups.
-                self.logger.error(f'No such group "{self.context.group}".')
-                return 1
-
-        tool_not_found = True  # Not found, yet
-        for remote in remotes:
-            tg_dir_r = self.tool_group_dir / remote
-            if not tg_dir_r.exists():
-                self.logger.warn(
-                    'No remote host "%s" in group %s.', remote, self.context.group,
-                )
+        groups = self.context.group.split(",")
+        for group in groups:
+            if self.verify_tool_group(group) != 0:
+                self.logger.warn(f'No such group "{group}".')
+                errors = 1
                 continue
 
-            if self.context.name:
-                names = [self.context.name]
-            else:
-                # Discover all the tools registered for this remote
-                names = self.tools(tg_dir_r)
-                if not names:
-                    # FIXME:  this is another odd case -- the remote subdirectory
-                    #  exists, but it's empty.  (We'll remove it below.)
+            try:
+                remotes = self.context.remote.split(",")
+            except Exception:
+                # We were not given any remotes on the command line, build the list
+                # from the tools group directory.
+                remotes = self.remote(self.tool_group_dir)
+                if not remotes:
+                    # Unless it is not a default group we will remove any
+                    # empty tool directories lingering or wasn't cleaned before.
+                    try:
+                        if group != "default":
+                            shutil.rmtree(self.tool_group_dir)
+                    except OSError:
+                        self.logger.error(
+                            "Failed to remove empty group directory %s",
+                            self.tool_group_dir,
+                        )
+                        return 1
```

Comment:
In the latest revision, it looks like you've removed the `return` statement, which I think was a good idea in the former context, but now control just falls through, which is probably not the best.  I suppose, given that `remotes` is empty, it's a quick trip to the end of the function, but, now that this code is in a subroutine, I think it would be good and appropriate, actually, to put the `return` statement back in.  😁

I could be easily swayed either way, but I am inclined to think that we should be declaring an error in this case, no?

Reply:
Let's see, if the `remotes` is empty and the corresponding group is not `default` then we remove that directory and we continue (in this case we return 0, [] as `errors` and `tools_not_found`) from `_clear_remotes` function since we got no errors and the intended action was completed.

When the we can't remove the group directory, we are logging the error and then continue to return 0, [] (as `errors` and `tools_not_found`). This is where I think we can return the `errors` message as 1 instead of 0, but we don't want to exit the program, do we? because this may be an outlier case for one group but we may have more groups to process after it.

Reply:
Agreed:  we want this function to return a non-zero value for `errors`, and there's no reason not to do it immediately, and then we want the caller to note it and return a non-zero value once it is done with its loop.

However, this points out a bug in the current code:  the call to this function is _overwriting_ the value of `errors` in the caller!  That is, if there were errors noted on previous iterations and then this function returns success, then `errors` will be set back to zero and the previous state will be lost.

Reply:
Without giving it a great deal of thought, my knee-jerk reaction is that `errors` should be the equivalent of a boolean value (that is, even if it's not an actual boolean, it represents "success" by a zero value and "failure" by a non-zero value, which is effectively a boolean).  I'm generally against "coded" return values (because someone has to document what the codes mean, and the caller or human has to be able to acquire and/or use that information), so a "boolean" result is what I would recommend (although, I'm good with differentiating between "you screwed up the command line" vs "I couldn't do what you asked" via return codes, but that's not specific to any particular command).

However, the fact that you felt inclined to resort to a tuple for the return value, one component of which is an error flag, suggests that we may have missed the bus:  if you're trying to return **_either_** a produced value _or_ an error indicator, then you probably want to employ exceptions to indicate the error(s) -- that way, you don't need a fancy return value.  (Of course, if you don't care about the "identity" of the error, you also have the option of returning either the list or `None` to indicate failure...which would also be better than returning a tuple.)

The question is, do we have a case in which we want to return _both_ a list and in indication that there was an error?    For the case that we're considering, I think the answer is "no", but I think there are other places in the function which detect errors.  So, that implies that we would want to return _both_ items, which starts to make the tuple look attractive.  (We could defer raising the exception until the function is finished, but we would still want to return the list, and conveying a return value via an exception object is a bit hacky.)

So, I guess you just need to have the caller receive the callee's error indicator in a temporary variable, and then check it to determine whether to set the flag in the caller's `errors` variable.

(Or, did I misunderstand the question?  😀)